### PR TITLE
Formulas

### PR DESCRIPTION
**Pull request by timgl**
## Changes

Add formulas to trends. Will need to add some docs/explanations but good enough for now. Behind the `3275-formulas` feature flag.

Also went Clickhouse only as that means we can use their special features and it's a fairly advanced feature

image

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

File: ee/clickhouse/queries/trends/formula.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                    )
+                    for i, query in enumerate(queries[1:])
+                ]
+            )
+            if filter.breakdown
+            else "".join(
+                ["CROSS JOIN ({}) as sub_{}".format(query, letters[i + 1]) for i, query in enumerate(queries[1:])]
+            ),
+        )
+        print(format_sql(sql, {"formula": filter.formula}))
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/formula.py:44]: 🖨

File: ee/clickhouse/queries/test/test_trends.py [hunk 2]
```diff
@@ -285,3 +285,105 @@ def test_action_with_prop(self):
...
+                        ],
+                        "formula": "A / B",
+                    }
+                ),
+                self.team,
+            )
+        self.assertEqual(action_response[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 4.0, 0.0, 0.0])
+
+        # formula with breakdown
+        with freeze_time("2020-01-04T13:01:01Z"):
```
Comment (EDsCODE) [inline ee/clickhouse/queries/test/test_trends.py:371]: How should I be applying the formula to a breakdown query with 3 elements?

File: ee/clickhouse/queries/trends/clickhouse_trends.py [hunk 3]
```diff
@@ -18,3 +31,47 @@ def _set_default_dates(self, filter: Filter, team_id: int) -> Filter:
...
+            return self._run_formula_query(filter, team.pk)
+
+        result = []
+        for entity in filter.entities:
+            if entity.type == TREND_FILTER_TYPE_ACTIONS:
+                try:
+                    entity.name = actions.get(id=entity.id).name
+                except Action.DoesNotExist:
+                    continue
+            entities_list = handle_compare(filter, self._run_query, team, entity=entity)
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/clickhouse_trends.py:74]: Cumulative is never reached

File: ee/clickhouse/queries/trends/normal.py [hunk 4]
```diff
@@ -62,12 +62,11 @@ def _format_normal_query(self, entity: Entity, filter: Filter, team_id: int) ->
             agg_query = self._determine_single_aggregate_query(filter, entity)
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/normal.py:62]: I think this function's queries are missing the new `date` and `data` notations which is causing the query to fail when table/pie is requested

**Review by EDsCODE — COMMENTED** [review]
Did a QA and some extra comments pinpointing where the problems occur:
 
1. When I try to use current_url contains with 2 entities and a formula (A+B) the query fails
2. Should add a test when `compare: true`
3. Should clear/disable formula when lifecycle is being viewed otherwise it returns error because lifecycle defaults to 1 element
4. Changing the chart to another chart type (table, pie) gives an error return (Error: `'sub_A.date' in table 'sub_A'`. 
5. Cumulative isn't cumulative. It mirrors linear chart type

It might be worth creating a new test class per feature on insight page and testing it against all the abstract class stubs. If we do this for all existing functionality on insights page, we would have very good coverage.

**Comment by timgl** [comment]
@EDsCODE Added a bunch of tests for the various cases.

**Review by EDsCODE — COMMENTED** [review]
Quick QA to double check some conditions

1. Filtering and breaking down by current_url specifically breaks
Screen Shot 2021-02-11 at 1 01 13 PM

2. Lifecycle disables formula but seems to still be applying it which blanks out
Screen Shot 2021-02-11 at 1 02 05 PM

3. If you do 2 entites, and A+B formula, then select table as chart type, it crashes the page

**Comment by timgl** [comment]
@EDsCODE fixed points 2 and 3, still can't reproduce point 1 though. any errors?

**Pull request merged**